### PR TITLE
Implement "list of options" concept on `HTMLSelectElement`.

### LINKED ISSUE #13763: Implement "list of options" concept on HTMLSelectElement
https://html.spec.whatwg.org/#concept-select-option-list

Add a new `list_of_options` (or maybe just `options`) function on [`HTMLSelectElement`](https://github.com/servo/servo/blob/master/components/script/dom/htmlselectelement.rs) and use it everywhere appropriate in that file [instead of doing things like this](https://github.com/servo/servo/blob/a80abf6c1dd0f14b40499d55129a53d397c3a3cd/components/script/dom/htmlselectelement.rs#L97).

### PR DESCRIPTION
**Pull request by frewsxcv**
Fixes #13763.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14085)

### CODE REVIEW

File: components/script/dom/htmlselectelement.rs [hunk 1]
```diff
@@ -385,3 +381,67 @@ impl VirtualMethods for HTMLSelectElement {
...
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (ret, curr, in_group) = find_next_option(&self.curr, self.in_group);
+        self.curr = curr;
+        self.in_group = in_group;
+        ret
+    }
+}
+
+fn find_next_option(curr: &Option<Root<Node>>, in_group: bool) -> (Option<Root<HTMLOptionElement>>, Option<Root<Node>>, bool) {
```
Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: If anyone has any better ideas for implementing this logic, please let me know!

Comment (nox) [inline components/script/dom/htmlselectelement.rs]: Can't you just keep a single HTMLOptionElement in the iterator and decide whether to go up or down the tree depending on whether its parent is an HTMLSelectElement or an HTMLOptGroupElement?

Comment (frewsxcv) [inline components/script/dom/htmlselectelement.rs]: I went back and forth. See the latest commit for those changes.

**Comment by frewsxcv** [comment]
Rebased. @nox Did you ever have any thoughts about this?

File: tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<!doctype html>
+<meta charset="utf-8">
+<title>HTMLSelectElement.value</title>
+<link rel="help" href="https://html.spec.whatwg.org/multipage/forms.html#dom-select-value">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<div id=log></div>
+
+<select id=sel1>
```
Comment (asajeffrey) [inline tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html]: Include a test for a select with no options?

File: tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  assert_equals(select.value, '1');
+});
+
+test(function() {
+  var select = document.getElementById('sel2');
+  assert_equals(select.value, '1');
+});
+
+test(function() {
+  var select = document.getElementById('sel3');
```
Comment (asajeffrey) [inline tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html]: Test for some failing gets?

File: components/script/dom/htmlselectelement.rs [hunk 4]
```diff
@@ -84,10 +84,16 @@ impl HTMLSelectElement {
                            HTMLSelectElementBinding::Wrap)
     }
 
+    // https://html.spec.whatwg.org/multipage/#concept-select-option-list
+    fn list_of_options(&self) -> impl Iterator<Item=Root<HTMLOptionElement>> {
+        ListOfOptions {
```
Comment (asajeffrey) [inline components/script/dom/htmlselectelement.rs]: That's annoying. We could work around that with a type
```rust
   enum Choice3<I,J,K> { First(I), Second(J), Third(K) }
   impl<I,J,K,T> Iterator<Item=T> for Choice3<I,J,K>
      where I: Iterator<Item=T>, J: Iterator<Item=T>, K: Iterator<Item=T>
   { ... }
```

File: tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html (hunk 3 above)
Comment (frewsxcv) [inline tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html]: What would that be testing with regards to this functionality?

**Comment by frewsxcv** [comment]
Comments have been addressed.

File: components/script/dom/htmlselectelement.rs [hunk 5]
```diff
@@ -84,10 +85,25 @@ impl HTMLSelectElement {
...
+                    let node = Root::downcast::<HTMLOptionElement>(node).unwrap();
+                    Choice3::First(iter::once(node))
+                } else if node.is::<HTMLOptGroupElement>() {
+                    Choice3::Second(node.children().filter_map(Root::downcast))
+                } else {
+                    Choice3::Third(iter::empty())
+                }
+            })
+    }
+
```
Comment (asajeffrey) [inline components/script/dom/htmlselectelement.rs]: Yay, much nicer. Pity the `Choice3` isn't pre-canned, but still yay!

File: components/script/dom/htmlselectelement.rs [hunk 6]
```diff
@@ -394,3 +399,23 @@ impl Validatable for HTMLSelectElement {
         true
     }
 }
+
+enum Choice3<I, J, K> {
+    First(I),
+    Second(J),
+    Third(K)
```
Comment (asajeffrey) [inline components/script/dom/htmlselectelement.rs]: Nit: put a comma on the end of this line?

File: tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  var div = document.createElement('div');
+  select.appendChild(div);
+  div.appendChild(option);
+  assert_equals(select.value, '');
+}, 'option is child of div');
+
+test(function() {
+  var select = document.getElementById('sel4');
+  assert_equals(select.value, '');
+}, 'no options');
```
Comment (asajeffrey) [inline tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html]: Corner case tests too.

**Review by asajeffrey — APPROVED** [review]
Looks good to me, squash and r=me.

File: tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html (hunk 7 above)
Comment (frewsxcv) [inline tests/wpt/web-platform-tests/html/semantics/forms/the-select-element/select-value.html]: Which corner cases are you referring to?

**Comment by KiChjang** [comment]
@bors-servo retry
* Let's rock mac-rel-wpt2

**Comment by larsbergstrom** [comment]
@bors-servo retry
- Machine started running jobs before fully provisioned

**Comment by KiChjang** [comment]
@bors-servo retry
* Looks like arm32 is also affected by the intermittent git issues

**Comment by KiChjang** [comment]
@bors-servo retry
* The annoying git

**Comment by KiChjang** [comment]
@bors-servo retry
* The super annoying git

**Comment by KiChjang** [comment]
@bors-servo retry
* The extremely annoying git

**Comment by KiChjang** [comment]
@bors-servo retry
* Same old network problem

**Comment by KiChjang** [comment]
@bors-servo retry
* No you don't

**Comment by KiChjang** [comment]
@bors-servo retry
* For real this time

**Pull request merged**